### PR TITLE
enhanced session recording integration tests

### PR DESCRIPTION
**Pull request by capnspacehook**
The current tests in `lib/bpf` don't sufficiently exercise our bpf tracing code. I could add to them, but I think it's valuable now and especially in the future to test ESR at a higher level that's closer to how ESR functions normally. I wanted to test that expected events are emitted when monitored processes are created using our full reexec logic in `lib/srv`, as current tests reexec and place child processes in a cgroup manually. If and when we change our reexec logic, the bpf tests will need to be updated.

To do this, I either had to create a full integration test in `integration`, or leverage the mocks and helpers already in `lib/srv` to reexec child processes. I chose the latter, as for testing ESR the only thing we're worried about is that we emit correct events based off of what child processes of an SSH session are doing. And the important process creation code in `lib/srv` isn't mocked, only the stuff we don't care about when testing ESR: SSH connections, an SSH server, etc.

My goal with these integration tests was to make it easy to add new test cases, and to streamline debugging failing tests. I tried to be as thorough as possible testing edge cases I know of, I think this is a good start at least. Feel free to suggest additional test cases if you think they'd be useful.

In the future I may remove some tests in `lib/bpf` that are redundant, but I'd like this PR to be mostly additive for ease of backporting later.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -453,7 +453,33 @@ tctl-app:
...
+		echo "bpf tests must be run as root"; \
+		exit 1; \
+	fi
+	TELEPORT_BPF_TEST=1 _test/libbpf.test -test.run=Test
+	TELEPORT_BPF_TEST=1 _test/libsrv.test -test.run=TestBPF
+
+# BPF tests will not work in a docker container and so should not be
+# run in CI for now.
+.PHONEY: test-bpf
+test-bpf: build-bpf-tests run-bpf-tests
```
Comment (codingllama) [inline Makefile]: Suggestions: use mktemp, use sudo, simplify all to one target. (It could also be script if it gets too complex.)

```suggestion
test-bpf: D=$(shell mktemp -dt 'teleport-bpf')
test-bpf:
	go test -c -tags bpf -o "$(D)/libbpf.test" ./lib/bpf
	go test -c -tags bpf -o "$(D)/libsrv.test" ./lib/srv
	TELEPORT_BPF_TEST=1 sudo "$(D)/libbpf.test" -test.run=Test
	TELEPORT_BPF_TEST=1 sudo "$(D)/libsrv.test" -test.run=TestBPF
```

File: Makefile [hunk 2]
```diff
@@ -453,7 +453,33 @@ tctl-app:
...
+# module and build cache of the current user but run the tests as root
+# as is required.
+.PHONEY: run-bpf-tests
+run-bpf-tests:
+	@if [ "$$(id -u)" -ne 0 ]; then \
+		echo "bpf tests must be run as root"; \
+		exit 1; \
+	fi
+	TELEPORT_BPF_TEST=1 _test/libbpf.test -test.run=Test
+	TELEPORT_BPF_TEST=1 _test/libsrv.test -test.run=TestBPF
```
Comment (codingllama) [inline Makefile]: Do we need to filter the test with -test.run? I'm afraid this could lead us to not running new tests in the future.

Maybe skip the -test.run filter for the lib/bpf package?

(Applies to the comment on test-bpf.)

File: lib/srv/bpf_test.go [hunk 3]
```diff
@@ -0,0 +1,850 @@
...
+				{
+					cmdInfo: commandInfo{
+						program: "ls",
+						args:    []string{"-la", cmdDir},
+					},
+					paths: []string{
+						"/etc/nsswitch.conf",
+						"/etc/passwd",
+						"/etc/group",
+						"/etc/localtime",
```
Comment (codingllama) [inline lib/srv/bpf_test.go:281]: I get cmdDir, but why are the /etc files on paths?

File: lib/srv/bpf_test.go [hunk 4]
```diff
@@ -0,0 +1,850 @@
...
+			command: fmt.Sprintf("find %s", cmdDir),
+			eventInfos: []expectedEvents{
+				{
+					cmdInfo: commandInfo{
+						program: "find",
+						args:    []string{cmdDir},
+					},
+					paths: []string{
+						"/proc/filesystems",
+						".",
```
Comment (codingllama) [inline lib/srv/bpf_test.go:216]: Why /proc/ and . ?

File: lib/srv/bpf_test.go [hunk 5]
```diff
@@ -0,0 +1,850 @@
...
+type addrInfo struct {
+	// destination address the command will connect to
+	addr string
+	// destination port the command will connect to
+	port int
+}
+
+// TestBPFRecording runs various commands with Enhanced Session Recording
+// enabled and verifies that the recorded events are expected.
+func TestBPFRecording(t *testing.T) {
```
Comment (codingllama) [inline lib/srv/bpf_test.go:117]: Apart from BPF does this test expect a certain environment? For example, a certain container/distro, unusual binaries, specific directory contents, etc?

File: lib/srv/bpf_test.go [hunk 6]
```diff
@@ -0,0 +1,850 @@
...
+					cmdInfo: commandInfo{
+						program: "cat",
+						// argv[0] is counted, so we expect MAXARGS-1 args
+						args: slices.Repeat([]string{tempFilePath}, maxArgs-1),
+					},
+					paths: []string{tempFilePath},
+				},
+			},
+		},
+		// TODO: bpf C code seems to want to add '...' if arguments are truncated but doesn't
```
Comment (codingllama) [inline lib/srv/bpf_test.go]:
```suggestion
		// TODO(capnspacehook): bpf C code seems to want to add '...' if arguments
		// are truncated but doesn't.
```

File: lib/srv/bpf_test.go [hunk 7]
```diff
@@ -0,0 +1,850 @@
...
+	}()
+
+	// Program should have executed now. If the complete signal has not come
+	// over the context, something failed.
+	select {
+	case <-time.After(5 * time.Second):
+		// We're not interested in the error, we just want to clean up the
+		// process.
+		_ = scx.killShellw.Close()
+		t.Fatalf("Timed out waiting for process to finish.")
```
Comment (codingllama) [inline lib/srv/bpf_test.go]: nit:

```suggestion
		t.Fatal("Timed out waiting for process to finish.")
```

File: bpf/README.md [hunk 8]
```diff
@@ -16,17 +16,35 @@ This document will help you get started.
...
 ```
 
+You can build BPF files faster using your local clang compiler using `make bpf-bytecode`.
+You will still need to run `make -C build.assets bpf-bytecode` to create reproducible files,
+otherwise the CI job that runs `make bpf-up-to-date` will fail.
+
 Note that there have been some issues with command tracing reading commands and arguments in a arm64 VM. You may
 need to test/develop on an EC2 instance instead.
 
+## Testing
```
Comment (codingllama) [inline bpf/README.md:34]: Any instructions for running on macOS?

**Review by codingllama — APPROVED** [review]

File: bpf/README.md (hunk 8 above)
Comment (capnspacehook) [reply bpf/README.md:34]: ESR code is Linux specific, so you'd have to either run on a VM (but note limitations just above) or on a different machine, ie an EC2 instance.

File: lib/srv/bpf_test.go (hunk 5 above)
Comment (capnspacehook) [reply lib/srv/bpf_test.go:117]: I tried to only use binaries from the coreutils package for the most part, `curl` is an obvious exception. I'll make the errors clearer when a needed command isn't found.

We will likely be running these tests on EC2 instances in the future as you can't load bpf programs in a container so they don't work in CI like normal tests. So in that case we'll be easily able to control what's installed.

File: Makefile (hunk 2 above)
Comment (capnspacehook) [reply Makefile]: I'll try again but I'm pretty sure I ran into some tests not running when I didn't specified `-run Test`, which in hindsight if true is probably a Go bug. That's kinda unlikely though so I'll confirm.

File: Makefile (hunk 1 above)
Comment (capnspacehook) [reply Makefile]: I was operating under the assumption that we didn't want to use `sudo` in our makefiles, but that would be preferable

File: lib/srv/bpf_test.go (hunk 3 above)
Comment (capnspacehook) [reply lib/srv/bpf_test.go:281]: These are files `ls` always opens when certain flags are passed. It needs user, group, time, etc information to print in the long format with `-l`. As far as I can tell these are Filesystem Hierarchy Standard (FHS) paths as well, so should generally work on Linux systems. An easy way to assert more expected disk events were emitted.

`/etc/passwd` and `/etc/group` are mentioned in the FHS 3.0 spec directly: https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.txt
`/etc/localtime` is used by glibc and the `localtime` manpage state's it should be there: https://www.man7.org/linux/man-pages/man5/localtime.5.html
Same with `/etc/nsswitch.conf`: https://man7.org/linux/man-pages/man5/nsswitch.conf.5.html

File: lib/srv/bpf_test.go (hunk 4 above)
Comment (capnspacehook) [reply lib/srv/bpf_test.go:216]: See above comment, `ls` and other commands always seem to open them. Tested on 2 different machines with different Linux distros. If this ever does become a reason the tests fail (I think it's very unlikely) they can be very easily removed.

File: Makefile (hunk 2 above)
Comment (capnspacehook) [reply Makefile]: Ok I'm dumb I just forgot that when tests are run in parallel the report when `-v` is passed probably will not include all tests. Will update this when I decide how the Makefile should be updated.

File: Makefile (hunk 1 above)
Comment (capnspacehook) [reply Makefile]: I don't think using `mktemp` is a good idea though, the idea is to build the test binaries in the same place each time so that if the test doesn't have to be re-compiled the test binary can be re-used. Otherwise the test binary will need to be linked every time, even if no code changes.

Comment (codingllama) [reply Makefile]: I don't know if we have sudo restrictions, but given the intended usage I think it's fine here.

Re mktemp: it should still use the Go build cache, so if the tradeoff is spurious files in the source code tree vs one rebuild per boot, I'll take latter. IMO it's good enough. It's just a suggestion, so resolve however you prefer.

File: lib/srv/bpf_test.go (hunk 3 above)
Comment (codingllama) [reply lib/srv/bpf_test.go:281]: Thanks for the thorough reply! Makes sense.

Comment (codingllama) [reply lib/srv/bpf_test.go:281]: It could be worth leaving a brief comment in there, but up to you.

File: lib/srv/bpf_test.go (hunk 4 above)
Comment (codingllama) [reply lib/srv/bpf_test.go:216]: Thanks! Same suggestion as above.

File: lib/srv/bpf_test.go (hunk 5 above)
Comment (codingllama) [reply lib/srv/bpf_test.go:117]: > I'll make the errors clearer when a needed command isn't found.

Sounds good!

File: bpf/README.md (hunk 8 above)
Comment (codingllama) [reply bpf/README.md:34]: I'm curious how you are running it. I do have a lima setup, that is probably what I would try to use if needed.

**Review by codingllama — APPROVED** [review]

File: bpf/README.md (hunk 8 above)
Comment (capnspacehook) [reply bpf/README.md:34]: I mostly develop on a desktop tbh, when I need to I use the previously issued Thinkpad as well. A lot easier to dev on Linux hardware rather than in an EC2 instance

Comment (codingllama) [reply bpf/README.md:34]: I think I did add this at some point as well: https://github.com/gravitational/teleport/blob/master/bpf/Vagrantfile.

File: bpf/README.md [hunk 9]
```diff
@@ -16,17 +16,35 @@ This document will help you get started.
...
+
+Tests are split across two packages:
+
+* `lib/bpf` - unit tests mainly for BPF related Go code
+* `lib/srv` - integration tests that execute commands as Teleport SSH normally does and checks emmitted events
+
+Tests need to be run as root, and will be skipped unless the `TELEPORT_BPF_TEST` environment variable is set.
```
Comment (rosstimothy) [inline bpf/README.md]: Suggestion: Indicate which value the env should be set to. Some flags are parsed using strconv.Parse, others check that value isn't empty. In either case you will be saving the reader of this from having to find this env in the code to determine which values will enable the test.
```suggestion
Tests need to be run as root, and will be skipped unless the `TELEPORT_BPF_TEST=yes` environment variable is set.
```

File: lib/bpf/bpf_test.go [hunk 10]
```diff
@@ -428,6 +429,8 @@ func TestRootPrograms(t *testing.T) {
 
 // TestRootLargeCommands given commands with higher amount of characters
 // (length), ensure the command events are generated correctly.
+// Still useful to keep around, don't think integration tests can exercise
+// a long arg as argv[0] (the program).
```
Comment (rosstimothy) [inline lib/bpf/bpf_test.go]:
```suggestion
// Overlaps with coverage from unit tests because integrations tests cannot exercise a long arg as argv[0] (the program).
```

File: lib/srv/bpf_test.go [hunk 11]
```diff
@@ -0,0 +1,850 @@
...
+
+type addrInfo struct {
+	// destination address the command will connect to
+	addr string
+	// destination port the command will connect to
+	port int
+}
+
+// TestBPFRecording runs various commands with Enhanced Session Recording
+// enabled and verifies that the recorded events are expected.
```
Comment (rosstimothy) [inline lib/srv/bpf_test.go]:
```suggestion
// enabled and verifies that the recorded events appear in the audit log.
```

File: lib/srv/bpf_test.go [hunk 12]
```diff
@@ -0,0 +1,850 @@
...
+		command    string
+		eventInfos []expectedEvents
+	}{
+		{
+			name:    "no commands",
+			command: "true",
+		},
+		{
+			name:    "basic command",
+			command: fmt.Sprintf("find %s", cmdDir),
```
Comment (rosstimothy) [inline lib/srv/bpf_test.go]: Suggestion: Here and below prefer string concatenation over fmt.Sprintf for simple cases. See https://google.github.io/styleguide/go/best-practices#string-concatenation.
```suggestion
			command: "find " + cmdDir,
```

File: lib/srv/bpf_test.go [hunk 13]
```diff
@@ -0,0 +1,850 @@
...
+	})
+
+	return bpfSrv
+}
+
+func handleConnections(l net.Listener) {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			continue
```
Comment (rosstimothy) [inline lib/srv/bpf_test.go:678]: Are there any cases for which we should return instead of continue?

File: Makefile (hunk 1 above)
Comment (capnspacehook) [reply Makefile]: Revised to use sudo in one target.

File: lib/srv/bpf_test.go (hunk 3 above)
Comment (capnspacehook) [reply lib/srv/bpf_test.go:281]: Added comment to describe why these files are added.

File: lib/srv/bpf_test.go (hunk 13 above)
Comment (capnspacehook) [reply lib/srv/bpf_test.go:678]: I don't think so? The listener has already successfully created at this point, so I don't think it matters if the connection has been accepted cleanly or not. We just want to see network events, the connection doesn't have to be successful

File: bpf/README.md (hunk 9 above)
Comment (capnspacehook) [reply bpf/README.md]: Good point, done

File: lib/srv/bpf_test.go (hunk 12 above)
Comment (capnspacehook) [reply lib/srv/bpf_test.go]: Makes sense, done

File: lib/srv/bpf_test.go (hunk 13 above)
Comment (rosstimothy) [reply lib/srv/bpf_test.go:678]: What if the listener was closed?

Comment (capnspacehook) [reply lib/srv/bpf_test.go:678]: Ah nice catch. The listener is never closed, but the `wg.Wait` call in a cleanup func doesn't block because I accidentally created a goroutine in the function passed to `wg.Go`. Fixed now

File: lib/srv/bpf_test.go [hunk 14]
```diff
@@ -0,0 +1,871 @@
...
+
+func handleConnections(l net.Listener) {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				return
+			}
+		}
+		conn.Close()
```
Comment (rosstimothy) [inline lib/srv/bpf_test.go:680]: Don't we still need the continue for all other error types? Wouldn't we panic on the conn.Close call below otherwise?
```suggestion
		conn, err := l.Accept()
		if err != nil {
			if errors.Is(err, net.ErrClosed) {
				return
			}
			continue
		}
		conn.Close()
```

Comment (capnspacehook) [reply lib/srv/bpf_test.go:680]: We do need the continue you're right; surprisingly it doesn't panic without it though, I guess no other errors are returned normally

**Comment by capnspacehook** [comment]
Squashed changes into one commit.

**Review by rosstimothy — APPROVED** [review]
Do you plan to backport these tests?

**Comment by capnspacehook** [comment]
Most likely yes
